frontend  - redux-saga implementation + unit testing

ptitdam2001 opened the pull request:
- implementation of redux-saga
- usage of redux-saga for series list
- Integration of unit testing in frontend

Commit:
frontend : integration of saga into Series

Commit:
Merge branch 'develop' into feature/frontend-redux-saga-implementation

bat3 commented on client/web/src/features/series/serieSaga.ts:
```diff
@@ -0,0 +1,21 @@
+  yield takeLatest(actionTypes.FETCH_ALL_SERIES, fetchSeries)
+}
+
+export default function* userSaga() {
```
should be seriesSaga

bat3 commented on client/web/src/features/series/serieSaga.ts:
```diff
@@ -0,0 +1,21 @@
+import SerieService from './serieService'
+
+function* fetchSeries() {
+  const { data: series } = yield call([SerieService, 'getAll'])
```
```suggestion
  const { data: series } = yield call(SerieService.getAll()])
```
As in userSaga

ptitdam2001 replied:
it is not the good usage, it does not work

bat3 commented on client/web/src/features/series/actionTypes.ts:
```diff
@@ -0,0 +1,4 @@
+export const SET_CURRENT_SERIE = 'SET_CURRENT_SERIE'
```
```suggestion
export const SET_CURRENT_SERIE = 'SET_CURRENT_SERIES'
```
'Serie' doesn't exist in English

bat3 commented on client/web/src/features/series/serieReducer.ts:
```diff
@@ -0,0 +1,24 @@
+
+const initialState = {
+  list: [],
+  serie: null,
```
'Serie' doesn't exist in English

bat3 commented on client/web/src/models/serie.ts:
```diff
@@ -0,0 +1,4 @@
+export interface ISerie {
```
'Serie' doesn't exist in English

Need to check all files ;-)

bat3 commented on client/web/src/features/user/userService.ts:
```diff
@@ -0,0 +1,50 @@
+import { IUser } from '../../models/user'
```
You don't think we can create a crudService by setting only the type (Serie or User) and the base URL ?
It seems to be same code between userService and seriesService.

ptitdam2001 replied:
I don't think because we will have specific methods (for example login method)

ptitdam2001 replied:
it is that why I implemented one interface ICrudService

Commit:
Implement unit tests with jest on Series

Commit:
Frontend code review feedback

ptitdam2001 commented:
I finished the development for that branch, I am waiting for yours feedbacks. It may be complicated for non front developers.

To summary, I implemented some unit tests with Jest and we cover all about series . The global coverage is 50% for the moment. Morever, I implemented redux-saga which it is a sort of event listener inside our front store. It manages side-effects : asynchronous things like data fetching and impure things like accessing the browser cache

bat3 commented on client/web/src/features/series/seriesSaga.ts:
```diff
@@ -0,0 +1,21 @@
+  yield takeLatest(actionTypes.FETCH_ALL_SERIES, fetchSeries)
+}
+
+export default function* serieSaga() {
```
```suggestion
export default function* seriesSaga() {
```

bat3 commented on client/web/src/features/series/seriesService.ts:
```diff
@@ -0,0 +1,38 @@
+import { ICrudService } from '../common/ICrudService'
+import { HttpRequest } from '../common/HttpRequest'
+
+export class SerieService implements ICrudService<ISeries> {
```
```suggestion
export class SeriesService implements ICrudService<ISeries> {
```

bat3 commented on client/web/src/features/user/userSaga.ts:
```diff
@@ -0,0 +1,20 @@
+import UserService from './userService'
+
+function* fetchUsers() {
+  const users = yield call(UserService.getAll)
```
Not consitente with seriesSaga:
`const { data: series } = yield call([serieService, 'getAll'])`

ptitdam2001 replied:
User feature is not up to date, I will create a new branch about user is starting

bat3 commented on client/web/src/features/series/seriesService.spec.ts:
```diff
@@ -0,0 +1,106 @@
+
+jest.mock('../common/HttpRequest')
+
+describe('SerieService', () => {
```
```suggestion
describe('SeriesService', () => {
```

bat3 commented on client/web/src/i18n.ts:
```diff
@@ -25,6 +25,7 @@ i18n
     interpolation: {
       escapeValue: false, // react already safes from xss
     },
+    // initImmediate: false,
```
To delete

Commit:
Frontend : code review feedback
